test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ActionOverloadingTests/Tests/ActionOverloadingQueryTests.cs

```diff
@@ -0,0 +1,205 @@
+                };
+
+                // Expected actions for an order line entity with composite ID (-10, -10)
+                var expectedActionsOnOrderLine = new List<(string, string)>
+                {
+                    (MetadataPrefix + retrieveProduct, $"OrderLines(OrderId=-10,ProductId=-10)/{retrieveProductName}")
+                };
+
+                // Verify the actions in the payload for each entity type
+                VerifyActionInEntityPayload("People(-1)", expectedActionsOnPerson, mimeType);
+                VerifyActionInEntityPayload("People(0)", expectedActionsOnEmployee, mimeType);
+                VerifyActionInEntityPayload("People(-7)", expectedActionsOnSpecialEmployee, mimeType);
+                VerifyActionInEntityPayload("People(1)", expectedActionsOnContractor, mimeType);
+                VerifyActionInEntityPayload("Products(-10)", expectedActionsOnProduct, mimeType);
+                VerifyActionInEntityPayload("OrderLines(OrderId=-10,ProductId=-10)", expectedActionsOnOrderLine, mimeType);
+            }
+        }
+
+        private void VerifyActionInEntityPayload(string queryUri, List<(string, string)> expectedActions, string acceptMimeType)
+        {
+            ODataMessageReaderSettings readerSettings = new ODataMessageReaderSettings();
+            var args = new DataServiceClientRequestMessageArgs(
+                "GET",
+                new Uri(_baseUri.AbsoluteUri + queryUri, UriKind.Absolute),
+                usePostTunneling: false,
+                new Dictionary<string, string>(),
+                HttpClientFactory);
+
+            var requestMessage = new HttpClientRequestMessage(args);
+            requestMessage.SetHeader("Accept", acceptMimeType);
+
+            var responseMessage = requestMessage.GetResponse();
+
+            ODataResource entry = null;
+            using (var messageReader = new ODataMessageReader(responseMessage, readerSettings, _model))
+            {
+                var reader = messageReader.CreateODataResourceReader();
+
+                while (reader.Read())
+                {
+                    if (reader.State == ODataReaderState.ResourceEnd)
+                    {
+                        entry = (ODataResource)reader.Item;
+                    }
+                }
+            }
+
+            // Verify the expected action metadata and target against the actual ODataActions in the ODataEntry.
+            Assert.Equal(expectedActions.Count, entry.Actions.Count());
+            foreach (var expected in expectedActions)
+            {
+                var actions = entry.Actions.Where(a => a.Metadata.AbsoluteUri == _baseUri + expected.Item1);
+                bool matched = false;
+                foreach (var action in actions)
+                {
+                    if (_baseUri + expected.Item2 == action.Target.AbsoluteUri)
+                    {
+                        matched = true;

```

Comment:
Nit: maybe break out of the inner loop once matched.
```suggestion
                        matched = true;
                        break;

```